sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-  """  
-  #print(dfa)
-  #in_main_model = set()
-  sev_sinks = set()
-  state = "0"
-  stlst = ["0"]
-  #print('This seq', sequence.split(" "))
-  for event in sequence.split(" "):
-      sym = event.split(":")[0]
-      sev = rev_smallmapping[sym.split('|')[0]]
-      
-      #print('curr symbol ', sym, 'state no.', dfa[state][sym]) 
-      
-      state = dfa[state][sym]
-      isred = 0
-      
-      if state != "":
-         isred = dfa[state[0]]["isred"]
-      #print(state)
-      #if state != "":
-      #if isred == 1:
-      #      in_main_model.add(state[0])
-      if state == "":
-          # return -1 for low sev symbols
-          #print(sev)
-          if len(str(sev)) >= 2:
-                #print('high-sev sink found', sev)
-                #print(stlst[-1], sinks[stlst[-1]], stlst)
+        command += ["--" + key + "=" + kwargs[key]]
+
+    result = subprocess.run(["FlexFringe/flexfringe"] + command + [args[0]], stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+    print(result.returncode, result.stdout, result.stderr)
+
+
+# Step 6.1: Load the resulting S-PDFA model
+def load_model(model_file):
+    """Wrapper to load resulting model json file
+
+    Keyword arguments:
+    model_file -- path to the json model file
+    """
+
+    # Because users can provide unescaped new lines breaking json conventions
+    #   in the labels, we are removing them from the label fields
+    with open(model_file) as fh:
+        model_data = fh.read()
+        model_data = re.sub(r'\"label\" : \"([^\n|]*)\n([^\n]*)\"', r'"label" : "\1 \2"', model_data)
+
+    model_data = model_data.replace('\n', '').replace(',,', ',')
+    model_data = re.sub(',+', ',', model_data)
+    machine = json.loads(model_data)
+
+    dfa = defaultdict(lambda: defaultdict(str))
+    for edge in machine["edges"]:
+        dfa[edge["source"]][edge["name"]] = edge["target"]
+
+    # Even though the properties might not be needed, the node has to be present in the dictionary (for sinks)
+    for entry in machine["nodes"]:
+        dfa[str(entry['id'])]["isred"] = entry['isred']
+        dfa[str(entry['id'])]["isblue"] = entry['isblue']
+        dfa[str(entry['id'])]["issink"] = entry['issink']
+
+    return dfa
+
+
+def traverse(dfa, sinks, sequence):
+    """Wrapper to traverse a given model with a string to create a state subsequence.
+
+    Keyword arguments:
+    dfa -- loaded main model
+    sinks -- loaded sinks model
+    sequence -- space-separated string to accept/reject in dfa
+    """
+    sev_sinks = set()
+    state = "0"
+    state_list = ["0"]
+    for event in sequence.split(" "):
+        sym = event.split(":")[0]  # TODO: remove in case multi is removed in `generate_traces` method
+        sev = rev_smallmapping[sym.split('|')[0]]
+        state = dfa[state][sym]
+        if state == "":
+            # Only keep IDs of medium- and high-severity sinks
+            if len(str(sev)) >= 2:
```

Comment:
This part here is not really correct. It sort of worked, not really in an intended way. As I have found out after quite some debugging (since I wrote an if-check here, as mentioned in the comment), defaultdict's in Python create a new element when it is not present in the dict and is accessed (which was the case here). This leads to `-1` being added to (global) `sinks` variable, which should not be there (otherwise states with ID -1 will sort of be sinks).

This was the issue for CCDC dataset. In the image you can see a (still reversed) sequence where `netDOS` follows `vulnD` (which should not be the case, see also my comment above about splitting subsequences, but let's assume it is). Because the ID is removed from `vuldD` (as it is low-sev node), the next transition is from state `-1` which of course cannot be correct as there are no nodes in S-PDFA with ID -1 (except for one dummy node, but that's not a problem here).

![issue-non-sev-sinks](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/13e84305-00fe-4fdf-9226-5a63571db8d4)

The states are essentially the same, except for this -1 ID on some states. As for solutions, the IDs can be stored in a separate list here (only for transitions, while the one with `-1`s can be returned) or remove ID later. In either case, I can handle it in a separate PR (depending on your reply), since it will chang the resulting AGs and I will not have a baseline to compare to for this PR

Reply:
Create an issue for this. Might be a potential bug.

Reply:
I have created an issue for this (with all the images)